components/klab.component.random/src/main/java/org/integratedmodelling/random/adapters/RecreationIDBAdapter.java

```diff
@@ -0,0 +1,202 @@
+            	
+            	Builder obuilder = builder.startObject(scope.getTargetName(), siteName, makeScale(urn, shape, scope));
+            
+            	// Add attributes to each recreation area like the name and id.
+            	for (Map.Entry<String, Object> entry : area.entrySet()) {
+            		if (entry.getKey()!="lat" || entry.getKey()!="lon" || entry.getKey()!="name") {
+            			obuilder.withMetadata(entry.getKey(), entry.getValue());
+            		}
+            	}
+            	obuilder.finishObject();
+            
+            }
+            
+        }
+	
+	}
+	
+    // TODO: There's a limit of 1000 entries hardcoded in the API. To make sure that we retrieve everything and with the ad hoc knowledge that 
+    // there's always less than 1000 recreation sites per US state, when there are N states passed as arguments create N different GET calls to 
+    // the API one per each state and assemble later all the responses. In the case no state parameter is specified default is retrieving from 
+    // the entire US territory, in that case have a list of all the states hardcoded and create one GET message per state. Eventually it would 
+    // be ideal to identify automatically the states involved and also reject responses out of the geographical scope of the context.
+    private final List<String> USA_STATES = Arrays.asList("AL", "AK", "AZ", "AR", "CA", "CO", "CT", "DE", "FL", "GA", "HI", "ID",
+            "IL", "IN", "IA", "KS", "KY", "LA", "ME", "MD", "MA", "MI", "MN", "MS", "MO", "MT", "NE", "NV", "NH", "NJ", "NM",
+            "NY", "NC", "ND", "OH", "OK", "OR", "PA", "RI", "SC", "SD", "TN", "TX", "UT", "VT", "VA", "WA", "WV", "WI", "WY");
+    private List<String> buildRecreationIDBInput(Map<String,String> parameters) {
+        ArrayList<String> query = new ArrayList<>();
+        List<String> states = parameters.containsKey(STATE) ? Arrays.asList(parameters.get(STATE).split(",")) : USA_STATES;
+        for(Map.Entry<String, String> entry : parameters.entrySet()) {
+            if (entry.getKey().equals(STATE)) {
+                continue;
+            }
+            query.add(entry.getKey()+"="+entry.getValue());
+        }
+        String finalQuery = String.join("&", query);
+        return states.stream().map(state -> finalQuery + "&" + STATE + "=" + state).toList();
+    }
+
+	
+	private IGeometry makeScale(Urn urn, IShape shape, IContextualizationScope scope) {
+        List<IExtent> extents = new ArrayList<>();
+        extents.add(scope.getContextSubject().getScale().getTime());
+        extents.add(shape);
+        return Scale.create(extents).asGeometry();
+    }
+	
+	@Override
+	public IResource getResource(String urn) {
+		Urn kurn = new Urn(urn);
+        ResourceReference ref = new ResourceReference();
+        ref.setUrn(urn.toString());
+        ref.setAdapterType(getName());
+        ref.setLocalName(kurn.getResourceId());
+        ref.setGeometry("#S2");
```

Comment:
[Magic string](https://dev.to/ruben_alapont/magic-numbers-and-magic-strings-its-time-to-talk-about-it-ci2l) what does `#S2` mean?

Reply:
Values such as `S0`, `S1`, `S2`, etc. refer to the number of spatial dimensions. They are well known by the modellers, so they are not that "magical". However, I see why it could be interesting to turn them into named constants.

Reply:
Actually it is the standard mathematical notation see https://en.wikipedia.org/wiki/N-sphere. Obscure for the uninitiated but certainly accurate.